rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+1) Assuming the merged works coming into the enricher are stored within some Elasticsearch index, the existing API code for querying on the archive path could be repurposed for this.
+
+2) Before coming into the enricher we could build and store a directed graph representation of the archives in some data store, with each node being a work and the edges encoding relations. We will discuss this further in the [Graph Store](#graph-store) section.
+
+In either case, denormalisation of works within the enricher would consist of:
+
+1. Fetching the relations for a given work
+2. Emitting a new work containing the relevant fields embedded (`parts`, `partOf`, `precededBy`, `succeededBy`)
+3. Traversing the relations and emitting works for these in turn.
+
+Part `3` here is required as we have two way relations encoded for particular fields: for example, if in the tree below there is a node added with path `A/1/Z`, we not only need to send this work downstream but also send work `A/1` (containing updated `parts`), and works `A/1/X` and `A/1/Y` (containing updated `succeededBy`).
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Generally, for the insertion, update or removal of any given node, we need to update its siblings, its parent, and the whole subtree containing its descendants. The worst case for this process is an update of the root node, in which case the whole archive needs to be denormalised, with the largest archives containing in the order of 1000 works.
+
+## Commonalities with Other Pipeline Stages
+
+The matcher and merger stages have similar underlying functionality to the enricher. Abstractly these stages involve:
+
+1) Receiving a work as input and applying domain knowledge to build a graph of connections to other works
+2) Using the input work and its graph to emit a new work at the output, including any required data from the relations
+
+This points to a potential universal solution within the pipeline, such as using a common service, library or data store to assist with some of these tasks. Additionally, there are potential similarities with upcoming concepts work, where there has been some investigation into how concepts would fit within a graph model.
+
+There are a few issues to bear in mind when considering how a shared solution would work:
+
+* Whilst stage 1 (currently the matcher) could potentially also involve analysis of archive relations, it is important as stated above that the denormalisation in the enricher takes place on a merged works to ensure data consistency. This indicates the merger and enricher stages being separate and accessing works from two separate works stores (unmerged and merged).
+
+* The matcher currently uses a custom DynamoDB graph implementation which was built specifically for the purpose. This was never meant to be a general solution and a rewrite would likely need to be involved in the building of some common system.
```

Comment:
Possibly add the silver lining that the dynamo implementation is struggling even now - it wasn't built for dealing with such big trees

Reply:
"It was never built to be a general solution"

I'm not sure that's true – it was never built to *scale*, but that's a bit different. IIRC, the code was written in the idea that we might eventually reused it.

Having not looked recently; is it the graph database or the locking that doesn't scale?

Reply:
OK good to know, I actually don't know too much about the internals but I think @jamieparkinson is under the impression that it is the locking code that doesn't scale well. Amazon introduced transactions to DynamoDB in November 2018, which might have rendered the custom locking unnecessary?

From my point of view there are a few further issues with the use of DynamoDB for this purpose, they are not necessarily all show stoppers but things to be aware of:

* You can't specify schema / constraints in Dynamo in the same way you could in a dedicated graph database or relational. For example relationships pointing to non existent objects can easily be prevented in Postgres / Neo4J., Neo4J easily allows specifying different types of relationships (so can represent matches and parents easily within the same store), etc

* Graph operations (traversal etc) and querying of relationships will be built in the graph databases instead of requiring custom solutions

Reply:
I spent a few days working on the matcher/Dynamo issues a couple of weeks ago - it appears to be mostly the locking that doesn't scale, and in a way that transactions don't help with unfortunately

Reply:
actually to elaborate on this, I don't think the matcher graph store would scale to this sort of thing. It is a special-purpose graph store designed for storing a large number of small graphs which are not connected to each other, and so traversing/querying/updating larger graphs would be very unwieldy

Reply:
yes, that's true. It wasn't built with the idea that it couldn't be replaced but definitely the model was a lot of small disconnected graphs. It would not scale for this problem